File: ios/Classes/Data/SdkAuthError.swift
//
//  SdkAuthError.swift
//  exponea
//

import Foundation

class SdkAuthError {
    let errorCode: String
    let customerIds: [String: String]

    init(errorCode: String, customerIds: [String: String] = [:]) {
        self.errorCode = errorCode
        self.customerIds = customerIds
    }

    init(_ data: [String: Any?]) throws {
        self.errorCode = try data.getRequired("errorCode")
        let ids: [String: Any?] = try data.getOptional("customerIds") ?? [:]
        self.customerIds = ids.reduce(into: [String: String]()) { result, item in
            let value = item.value
            if value is NSNull { return }
            result[item.key] = String(describing: value)
        }
    }

    func toMap() -> [String: Any?] {
        return [
            "errorCode": errorCode,
            "customerIds": customerIds,
        ]
    }
}

enum SdkAuthErrorReasonMapper {
    static func map(_ reason: Any) -> String {
        switch String(describing: reason) {
        case "expiredSoon":
            return "TOKEN_ABOUT_TO_EXPIRE"
        case "expired":
            return "TOKEN_EXPIRED"
        case "invalid":
            return "TOKEN_REJECTED"
        case "notProvided":
            return "TOKEN_NOT_PROVIDED"
        case "insufficient":
            return "TOKEN_INSUFFICIENT"
        default:
            return "TOKEN_REJECTED"
        }
    }
}

File: android/src/main/kotlin/com/exponea/data/SdkAuthError.kt
package com.exponea.data

import com.exponea.sdk.models.SdkAuthError as NativeSdkAuthError

data class SdkAuthError(
    val errorCode: String,
    val customerIds: Map<String, String> = emptyMap(),
) {
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "errorCode" to errorCode,
            "customerIds" to customerIds,
        )
    }
}

fun NativeSdkAuthError.toBridge(): SdkAuthError {
    val ids = customerIds.mapNotNull { (key, value) ->
        value?.let { key to it }
    }.toMap()
    return SdkAuthError(
        errorCode = errorCode.name,
        customerIds = ids,
    )
}
